datafusion/physical-expr/src/equivalence.rs

```diff
@@ -42,8 +42,65 @@ use indexmap::IndexMap;
 pub type EquivalenceClass = Vec<Arc<dyn PhysicalExpr>>;
 
 /// Stores the mapping between source expressions and target expressions for a
-/// projection. Indices in the vector corresponds to the indices after projection.
-pub type ProjectionMapping = Vec<(Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>)>;
+/// projection.
+#[derive(Debug, Clone)]
+pub struct ProjectionMapping {
+    /// `(source expression)` --> `(target expression)`
+    /// Indices in the vector corresponds to the indices after projection.
+    inner: Vec<(Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>)>,
+}
+
+impl ProjectionMapping {
+    /// Constructs the mapping between a projection's input and output
```

Comment:
I tried to add some additional details here in the comments to make it easier to understand. The logic is the same